### PR TITLE
Allow using different listing strategies

### PR DESCRIPTION
**Pull request by fpetkovski**
In https://github.com/thanos-io/thanos/pull/6474 we changed the way blocks are discovered in object storage by replacing a top-level iteration + exist calls with a single recursive iteration. This method seems to be slow for some users who either have a large amount of objects in the bucket, or prefer faster boot times.

To resolve that problem, this PR exposes a flag allowing users to control the strategy used for discovering TSDB blocks. The default strategy will be one prior to #6474 since it has been there for a long time and has proven to work well for most casesl.

* [x] I added CHANGELOG entry for this change.
* [ ] Change is not relevant to the end user.

### CODE REVIEW

File: docs/components/compact.md [hunk 1]
```diff
@@ -279,6 +279,15 @@ usage: thanos compact [<flags>]
 Continuously compacts blocks in an object store bucket.
 
 Flags:
+      --block-discovery-strategy="concurrent"
+                                One ofconcurrent, recursive. When set to
```
Comment (mfoldenyi) [inline docs/components/compact.md:283]: nit. missing space

File: cmd/thanos/store.go [hunk 2]
```diff
@@ -137,6 +146,10 @@ func (sc *storeConfig) registerFlag(cmd extkingpin.FlagClause) {
 	cmd.Flag("sync-block-duration", "Repeat interval for syncing the blocks between local and remote view.").
 		Default("15m").DurationVar(&sc.syncInterval)
 
+	strategies := strings.Join([]string{string(concurrentDiscovery), string(recursiveDiscovery)}, ", ")
+	cmd.Flag("block-discovery-strategy", "One of"+strategies+". When set to concurrent, stores will concurrently issue one call per directory to discover active blocks in the bucket. The recursive strategy iterates through all objects in the bucket, recursively traversing into each directory. This avoids N+1 calls at the expense of having slower bucket iterations.").
```
Comment (mfoldenyi) [inline cmd/thanos/store.go:150]: "One of "+strategies
instead of
"One of"+strategies

File: docs/components/compact.md (hunk 1 above)
Comment (fpetkovski) [reply docs/components/compact.md]: Good catch, thanks 👍

File: cmd/thanos/store.go (hunk 2 above)
Comment (fpetkovski) [reply cmd/thanos/store.go]: Thanks, updated.

File: cmd/thanos/compact.go [hunk 3]
```diff
@@ -754,6 +763,9 @@ func (cc *compactConfig) registerFlag(cmd extkingpin.FlagClause) {
 		"as querying long time ranges without non-downsampled data is not efficient and useful e.g it is not possible to render all samples for a human eye anyway").
 		Default("false").BoolVar(&cc.disableDownsampling)
 
+	strategies := strings.Join([]string{string(concurrentDiscovery), string(recursiveDiscovery)}, ", ")
+	cmd.Flag("block-discovery-strategy", "One of "+strategies+". When set to concurrent, stores will concurrently issue one call per directory to discover active blocks in the bucket. The recursive strategy iterates through all objects in the bucket, recursively traversing into each directory. This avoids N+1 calls at the expense of having slower bucket iterations.").
```
Comment (yeya24) [inline cmd/thanos/compact.go:767]: We revert the current behavior and change back to the old one? Is it worth a changelog?

Comment (mfoldenyi) [reply cmd/thanos/compact.go:767]: I am no authority on anything, but I would personally like to have a changelog entry on this. Seeing how it was mentioned in different threads how this is a blocker for upgrading Thanos itself for some, I think it is important enough.

Comment (fpetkovski) [reply cmd/thanos/compact.go:767]: Good call. I added a changelog entry, let me know if that's sufficient or we should elaborate further.

Comment (mfoldenyi) [reply cmd/thanos/compact.go:767]: I like it. Good enough for me.

**Review by mfoldenyi — APPROVED** [review]
For what its worth, Looks Good To Me. :)

**Review by yeya24 — APPROVED** [review]
Thanks. Merging now. 

Thanks for helping with the review! @mfoldenyi

**Pull request merged**